### PR TITLE
fix task cancelation handling in processor

### LINKED ISSUE #14: Inspector.CancelProcessing does not work
It appears that `Inspector.CancelProcessing` does not work, or something is required to get it to work

There are currently no tests provided for `CancelProcessing`, but there are tests for `PublishCancelation` that it directly calls, but no issues seem to occur for those tests, though I have not looked into what those tests do in particular

I have added my own test in branch [`cancel-testing`](https://github.com/qluvio/asynq/commit/067548c102455e60b327f34dd3024fbac14a00c0), which currently fails. There is also cancel tests in `content-fabric` that try to use `CancelProcessing` but fail

It may be worth investigating further what the issue is, though we anticipate having to change the cancelation logic anyway, given work to remove waiting on channels in the processor

These open `asynq` issues may also be relevant, especially the first one
https://github.com/hibiken/asynq/issues/524
https://github.com/hibiken/asynq/issues/532

### PR DESCRIPTION
**Pull request by jslching**
Fixes #14

### CODE REVIEW

**Comment by elv-gilles** [comment]
Why not !

The only problem I see is that `resFn` in `processor.fini` will still be called another time (I think this was not the case before changes in #7) unless explicitly handled like below.

Maybe add a test around that.

```
func (p *processor) fini() {
			...
	resFn := func(t *processorTask) func(reflect.Value) {		
			...
			if resErr != nil {
				if resErr != t.ctx.Err() {
					// handled by ctxDoneFn
					p.handleFailedMessage(t.ctx, t.msg, "errored", resErr)
					t.cleanup()
				}
				return
			}
```

**Comment by jslching** [comment]
> The only problem I see is that `resFn` in `processor.fini` will still be called another time (I think this was not the case before changes in #7) unless explicitly handled like below.
> 
> Maybe add a test around that.

Hmmm, I see, I'll try adding a test to confirm, but I don't think this should happen. `reflect.Select` is single-threaded and receives from channels one at a time. Whenever it selects a `ctxDoneCh` or `resCh`, it will remove that channel and the corresponding paired channel (so a `ctxDoneCh` and a `resCh` are both removed), then it executes the `Fn` for the selected channel only. This way, if `ctxDoneFn` is called, `resFn` will not be called after, since the `resCh` would have been removed from the select cases before the next run of `reflect.Select`

Unless, of course, there is a bug :)

**Comment by jslching** [comment]
> Maybe add a test around that.

See `processor_test.go:TestProcessorContextDoneWithAsyncTasks()`

File: processor.go [hunk 1]
```diff
@@ -504,7 +504,7 @@ func (p *processor) handleFailedMessage(ctx context.Context, msg *base.TaskMessa
 		p.retry(ctx, msg, err, false /*isFailure*/)
 		return
 	}
-	if msg.Retried >= msg.Retry || errors.Is(err, SkipRetry) {
+	if msg.Retried >= msg.Retry || errors.Is(err, SkipRetry) || errors.Is(err, context.Canceled) {
 		p.logger.Warnf("Retry exhausted for task id=%s", msg.ID)
```
Comment (elv-gilles) [inline processor.go]: maybe - depending on case - change log message & level:

```suggestion
	if msg.Retried >= msg.Retry || errors.Is(err, SkipRetry) || errors.Is(err, context.Canceled) {
		if msg.Retried >= msg.Retry {
			p.logger.Warnf("Retry exhausted for task id=%s", msg.ID)
		} else {
			p.logger.Debugf("Archiving task id=%s (err:%v)", msg.ID, err)
		}
```

**Review by elv-gilles — APPROVED** [review]
ok, thanks - I must have been confused by the callbacks during my tests yesterday.

**Pull request merged**